Lab section for Visualize

thomasneirynck opened the pull request:
Continuation of https://github.com/elastic/kibana/pull/14197.

Changes:
- a new global-setting "lab-mode" is introduced. this is turned on by default.
- visualizations marked experimental, will only show up in the list if lab-mode is enabled

Changes wrt. 14197:
- merge experimental visualizations inside regular collection
- remove experimental vizzes from dashboard (if lab-mode is turned off)

@alexfrancoeur any suggestions for wording of the experimental-banner?

timroes commented on src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.js:
```diff
@@ -26,10 +27,21 @@ export class VisualizeEmbeddableFactory extends EmbeddableFactory {
+
+        if (!labMode && savedObject.vis.type.isExperimental) {
+          return new Embeddable({
+            title: `${savedObject.title} is an experimental visualization. 
```
I would rather not fit all of this in the title of the dashboard panel. This doesn't fit in the panel in most cases.
Could we rather write this in the body of the panel. My suggestion would be:

```js
if (!labMode && savedObject.vis.type.isExperimental) {
  domNode.innerHTML = `<div class="disabledLabVisualization">
      <div class="kuiVerticalRhythm disabledLabVisualization__icon kuiIcon fa-flask" aria-hidden="true"></div>
      <div class="kuiVerticalRhythm"><em>${savedObject.title}</em> is an experimental visualization.</div>
      <div class="kuiVerticalRhythm">Please turn on experimental mode in the advanced settings to see
      experiemental visualizations.</div>
    </div>`;
  return new Embeddable({
    title: savedObject.title
  });
}
```

and add the following to `src/core_plugins/kibana/public/visualize/styles/main.less` (or better create a `lab_vis.less` in the same folder and import that in the `main.less`):

```less
.disabledLabVisualization {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
}

.disabledLabVisualization__icon {
  font-size: 2em;
}
```

This would provide the following look and feel, which I think would be a bit more beautiful and scalable:

![screenshot-20171114-091249](https://user-images.githubusercontent.com/877229/32769404-0a865d34-c91c-11e7-9228-63f6fab4dadb.png)

timroes replied:
This above code of course still has the issue, that I would change the description (see in the comment below).

timroes commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing.js:
```diff
@@ -29,7 +29,10 @@ export function VisualizeListingController($injector) {
         this.listingLimit = config.get('savedObjects:listingLimit');
-        return result.hits;
+        return result.hits.filter(result => {
+          const labMode = config.get('visualize:labMode');
```
Move the `config.get` outside of the filter function so the option won't be loaded again for every result.

timroes commented on src/core_plugins/kibana/public/visualize/wizard/step_1.html:
```diff
@@ -15,9 +15,19 @@
+        Select visualization type
+      </h1>
+
+      <!--<div class="visualizeLabVisualizations" ng-if="labModeEnabled">-->
```
Before merging, remove all that commented code and the CSS classes not needed anymore.

timroes commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -78,6 +78,7 @@
           ng-blur="finder.hitBlur($event)"
           ng-click="finder.preventClick($event)">
           <span aria-hidden="true" class="finder-type fa" ng-if="hit.icon" ng-class="hit.icon"></span>
+          <div class="kuiIcon fa-flask ng-scope" ng-if="hit.type && hit.type.isExperimental"></div>
```
`ng-if="hit.type.isExperimental"` is enough. Angular will automatically safe traverse that object, and not fail if `hit.type` isn't defined.

thomasneirynck replied:
seriously? that's awesome! :)

timroes commented on src/core_plugins/kibana/ui_setting_defaults.js:
```diff
@@ -122,6 +122,10 @@ export function getUiSettingDefaults() {
       value: 100,
       description: 'Never show more than this many bars in date histograms, scale values if needed',
     },
+    'visualize:labMode': {
```
Why do we call it `labMode`? Also I think the description is pretty much saying nothing. "Enable lab mode in visualize" leaves the user with the question: what is lab mode? What will enable/disable this option actually cause?

I would rather rename the setting `visualize:enableExperimentalVis` and change the description to something like: "Enable experimental (not yet stable) visualizations in Kibana".

timroes commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -47,13 +47,27 @@ module.controller('VisualizeWizardStep1', function ($scope, $route, kbnUrl, time
 
   const visTypes = Private(VisTypesRegistryProvider);
+  $scope.labModeEnabled = config.get('visualize:labMode');
+  // $scope.isLabMode = $route.current.params.lab;
```
Remove commented code

timroes commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -47,13 +47,27 @@ module.controller('VisualizeWizardStep1', function ($scope, $route, kbnUrl, time
   kbnUrl.removeParam(DashboardConstants.ADD_VISUALIZATION_TO_DASHBOARD_MODE_PARAM);
 
   const visTypes = Private(VisTypesRegistryProvider);
+  $scope.labModeEnabled = config.get('visualize:labMode');
```
When removing the commented code in the template, this doesn't need to be in the `$scope` anymore and can be a local variable.

timroes commented on src/core_plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -47,13 +47,27 @@ module.controller('VisualizeWizardStep1', function ($scope, $route, kbnUrl, time
   const visTypes = Private(VisTypesRegistryProvider);
+  $scope.labModeEnabled = config.get('visualize:labMode');
+  // $scope.isLabMode = $route.current.params.lab;
+  $scope.toggleLabView = () => {
```
With the commented template this can be removed completely.

alexfrancoeur commented:
Great work! had a couple of feedback items as I walked through the PR. Let me preface this with the fact that I've had a control visualization persisted in Kibana for some time now, some of these issues I ran into may be related to that.

First, should we mark experimental visualizations that have been created with a beaker or text somehow on the visualization list views?

<img width="1131" alt="screen shot 2017-11-14 at 11 37 09 am" src="https://user-images.githubusercontent.com/25181823/32792058-6bf44982-c930-11e7-904d-51ca4f3c1572.png">

<img width="1113" alt="screen shot 2017-11-14 at 11 50 21 am" src="https://user-images.githubusercontent.com/25181823/32792704-0783a8a6-c932-11e7-8093-ad4c6f3fe78e.png">

It looks like we are from dashboards. Should we adopt the same format?

<img width="1220" alt="screen shot 2017-11-14 at 11 39 16 am" src="https://user-images.githubusercontent.com/25181823/32792095-87e95a42-c930-11e7-8348-f189b324e4cc.png">

Loving the banner!

<img width="1041" alt="screen shot 2017-11-14 at 11 41 07 am" src="https://user-images.githubusercontent.com/25181823/32792159-b96c4a5c-c930-11e7-995b-6adc4e3af32e.png">

 I have a suggestion for the description. This may be too blunt but let me know what you think. I borrowed a bit from our docs.

> This visualization is marked as experimental. The feature set is subject to change with each release and in non-backwards compatible ways if required. Have feedback? Please create an issue in [GitHub](https://github.com/elastic/kibana/issues/new).

I can't seem to disable lab mode when an experiment visualization has previously been created.
![nov-14-2017 11-44-26](https://user-images.githubusercontent.com/25181823/32792376-321a6cd6-c931-11e7-931e-340b11fc9638.gif)

But I'm presented with this error and can still see the experimental visualizations.

<img width="1277" alt="screen shot 2017-11-14 at 11 45 24 am" src="https://user-images.githubusercontent.com/25181823/32792425-54f8a70e-c931-11e7-9d89-131fecbac3ac.png">

This may be unrelated, but I'm not seeing a warning when attempting to delete the experimental visualization. I actually am unable to delete it. I am prompted with a warning in the saved object management view, but also cannot delete it.

<img width="918" alt="screen shot 2017-11-14 at 11 46 08 am" src="https://user-images.githubusercontent.com/25181823/32792503-7acec5a8-c931-11e7-9555-311f334e81b6.png">

I'm now stuck in this weird state where experimental mode is disabled but the new visualization cards and my old visualizations are still there. Given that I can't disable lab mode, I'm unable to see what the experience is like actually disabling it. How do we handle previously created experiment visualizations? Do we delete them? Do we hide them from the visualization list views and as an option to add to a dashboard? Do we leave them as?

alexfrancoeur commented:
Also, should we hint in the warning somewhere that you can turn this off under advanced settings?

alexfrancoeur commented:
Linking to https://github.com/elastic/kibana/issues/13951

thomasneirynck commented:
hi @alexfrancoeur ,

- good catch on the flask. I added it to the listing page
- also updated the description
- the warning seems to be unrelated, I can reproduce it in master
- for referring to the advanced settings, we do that in the new dashboard message, which is probably the place you want it (because it is the only plae where a user might run into a mismatch)

- the other error (readonly). I couldn't reproduce this. but ill look out for it.

timroes commented on src/core_plugins/kibana/public/visualize/listing/visualize_listing_table.js:
```diff
@@ -142,10 +142,21 @@ export class VisualizeListingTable extends Component {
+      <span>
+        {flaskHolder}
+        <a className="kuiLink" href={this.getUrlForItem(item)}>
+          {item.title} foobariontos
```
!["foobariontos" all our visualizations meme](https://user-images.githubusercontent.com/877229/32823624-23f341d4-c9de-11e7-943c-6998f81ebadd.jpg)

thomasneirynck replied:
ouch.... the sham

ppisljar commented:
and this is still missing the warning when saving a dashboard with experimental visualizations
should probably say something like:
"this dashboard contains experimental visualizations, which should not be used in production. are you sure you know what you are doing foobariantos?"

thomasneirynck commented:
@ppisljar imho, not sure if we need that extra dialog. That banner cannot be more clearer, and all listings are indicated with a flag. If we add an additional dialog with a warning upon save, I feel like we'd continue to pester the user.

thomasneirynck commented:
need to resolve merge conflicts. grown out of sync with master

ppisljar commented:
@thomasneirynck we don't need additional dialog, just additional message in the "save your dashboard" dialog. I think it helps to make it clear what's going on, as the biggest impact disabling this feature (in management, opting out of experimental visualizatons) is gonna be on the dashboard. (saved dashboards)

timroes commented on src/core_plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,6 +2,20 @@
+  flex: 1 1 auto;
+}
+
+.visualizeLabVisualizations {
```
This can be removed I guess?

thomasneirynck commented:
resolve merge conflict

thomasneirynck commented:
After some discussion  (@alexfrancoeur, @epixa), we decided to fine-tune this.

Some experimental features will be further along than others, and shouldn't be treated as lab-features. Correspondingly, the expectations for each should be clearly communicated.

I split this up, to make a distinction between lab-plugins and experimental plugins. I also slightly adjusted the wording in the info-banners for labs-plugins and experimental-plugins.

In practice, lab-features can be toggled on-off, but this doesn't apply to TSVB which remains experimental but we do not consider a lab-feature.

@alexfrancoeur @timroes can you take a second look? thx!

alexfrancoeur commented:
@thomasneirynck looks like the additional texts wraps over on a macbook. I imagine a lot of our users will be using a similar resolution. Mind getting rid of the `non-backwards compatible` bit and go back to what you originally had?

>This visualization is in labs mode. The feature set is subject to change with each release. Have feedback? Please create an issue in GitHub

<img width="1440" alt="screen shot 2017-11-16 at 9 02 07 pm" src="https://user-images.githubusercontent.com/25181823/32925760-94c5dd0c-cb11-11e7-955f-815c9c18646e.png">

Other then that, everything else works as expected.

<img width="1381" alt="screen shot 2017-11-16 at 9 08 14 pm" src="https://user-images.githubusercontent.com/25181823/32925881-4f269042-cb12-11e7-8d33-a45f928c8065.png">

thomasneirynck commented:
resolve merge conflicts with latest master

thomasneirynck commented:
hi @alexfrancoeur , these are the two banner-messages now:

labs mode:
> `This visualization is in labs mode. The feature set is subject to change with each release.
        Have feedback? Please create an issue in GitHub

tsvb:
> This visualization is marked as experimental. 
Have feedback? Please create an issue in GitHub

alexfrancoeur commented:
>Could we please end the warnings with a period? All sentences in the warning except the last one are accompanied by a punctuation mark.

+1 on this. We'll be making our content structure / formatting more strict moving forward. I agree with Tim her, let's throw a `.` at the end of those warnings.

>Thus I would rather not have two independent flags, which in the end can be confusing how those play together, but rather a flag (e.g. called stage), that can have either value of stable (default), experimental or lab.

This seems like more of an implementation detail but it seems to me this would be more future proof and allow for other stages such as `alpha` or `beta` as well if we ever decided to take this concept further. With 6.1 FF around the corner though, is this something we can address in the next minor?

>The "lab/experimental" icon in the vis lists is currently only shown for experimental, but not for lab visualizations.

The icons in most list views seem to be working for me for both experimental and labs

<img width="934" alt="screen shot 2017-11-17 at 9 28 18 am" src="https://user-images.githubusercontent.com/25181823/32951961-adfc77b4-cb79-11e7-84f8-30fdc1fc2027.png">

<img width="1368" alt="screen shot 2017-11-17 at 9 29 10 am" src="https://user-images.githubusercontent.com/25181823/32952099-2a99a63e-cb7a-11e7-8a45-8d1fb28c31d7.png">

Unless this is what you're referring to. I agree that explicitly calling out labs here would be useful. +1

<img width="256" alt="screen shot 2017-11-17 at 9 33 04 am" src="https://user-images.githubusercontent.com/25181823/32952155-5aa3f708-cb7a-11e7-96e4-fb54d05743ff.png">
<img width="299" alt="screen shot 2017-11-17 at 9 33 09 am" src="https://user-images.githubusercontent.com/25181823/32952168-6093b428-cb7a-11e7-8d22-84e9e2115a54.png">

timroes commented:
@alexfrancoeur they are just working in the lists, because all those plugins ALSO have the `isExperimental` applied. If you would remove this from them and just leave them in `isLab` mode no icon would be shown anymore... thus bringing us back to boolean flags maybe not being the best type in that case :D

thomasneirynck commented:
@timroes @alexfrancoeur I agree we can need to fix he wording in the tooltips, the shoddy punctuation, and the implementation around providing urls. I will make those changes.

imho, we shouldn't get too preoccupied with the taxanomy and naming of these categories of plugins; ie. What is "regular", "labs", "experimental". Are they mutually exclusive. Is one a superset of the other? Should we name them labs, experimental, alpha, beta? etc... There's good arguments for any sort of combination of those. I really don't know the right answer there. 

Apart from the information-architecture and the naming concerns, we do know what the practical outcomes should be, at least for 6.1.

- We need to set the stage for later minors, so we can do soft-intros of new plugins. This is especially important, because we know we want some opportunity to iterate for the kind of future feature we are looking at (vega, timelines, new UX, geo-app, ...).
- We need a way for users to turn or off the cutting-edge new features (call them labs, alpha, super-experimental, rough-diamonds, ....). That is what the advanced-setting enableLabs-flag is for.
- We need a special case for plugins like TSVB. We have a de facto commitment to this plugin, even though it is labeled experimental. Let's not ditch the experimental flag, it is a legacy naming. What's in a name after all. I didn't realize this early enough. So let's just treat it the same as we always have.
- We need some way to set expectation to the user. That is what the banner-messages are for. It needs to be accurate, but we cannot enumerate all our expectations there. We're still testing the waters of what labs-mode means.

So for 6.1, I'd like to see that we get those four outcomes right. I don't think we need to fix the icon-problem perfectly. We can always improve on the design and the icons. So right now, we don't use special tags for lab-features, that is OK. As for the taxonomy of how all these concepts relate, here too we have some flex on how this evolves. If we want to evolve this to a more rigid hierarchy, we always can later.

alexfrancoeur commented:
@thomasneirynck LGTM after we add a period to those warnings. I did not mean to call out the tooltips on a new viz, just was trying to figure out what @timroes was referencing. Though now I'm wondering if we should consider `rough-diamond-mode` 💎 instead of labs. (just kidding, we definitely shouldn't 😄 )

ppisljar commented:
jenkins, test this

ppisljar commented:
`fail: "visualize app tile map visualize app tile map chart should save with zoom level and load, take screenshot"`

jenkins, test this

ppisljar commented:
`fail: "visualize app tile map visualize app tile map chart should save with zoom level and load, take screenshot"`

jenkins, test this

timroes commented:
Jenkins, test this

ppisljar commented:
replacing with  #15050
